test/cpp/src/701-dram-scheduler.cc

```diff
@@ -0,0 +1,109 @@
+#include <catch.hpp>
+#include "mocks.hpp"
+#include "defaults.hpp"
+#include "dram_controller.h"    
+#include "champsim_constants.h"
+#include <algorithm>
+#include <cfenv>
+#include <cmath>
+
+#include <iostream>
+
+std::vector<uint64_t> dram_test(MEMORY_CONTROLLER* uut, std::vector<champsim::channel::request_type>* packet_stream, std::vector<uint64_t>* arriv_time)
+{
+    uut->current_cycle = 0;
+
+     auto ins_begin = std::begin(uut->channels[0].RQ);
+    //load requests into controller
+    std::transform(std::cbegin(*packet_stream), std::cend(*packet_stream), std::cbegin(*arriv_time), ins_begin, [](auto pkt, uint64_t cycle) {
+        DRAM_CHANNEL::request_type r_pkt = DRAM_CHANNEL::request_type{pkt};
+        r_pkt.forward_checked = false;
+        r_pkt.event_cycle = cycle;
+        return r_pkt;
+    });
+
+    //carry out operates, record request scheduling order
+    uint64_t completed_reqs = 0;
+    std::vector<bool> last_scheduled(packet_stream->size(),false);
+    std::vector<uint64_t> scheduled_order;
+
+    while (completed_reqs < std::size(*packet_stream))
+    {
+        //operate mem controller
+        uut->_operate();
+        
+        //get scheduled requests
+        std::vector<bool> next_scheduled{};
+        std::transform(std::begin(uut->channels[0].RQ), std::end(uut->channels[0].RQ), std::back_inserter(next_scheduled), [](const auto& entry) { return ((entry.has_value() && entry.value().scheduled) || !entry.has_value()); });
+        
+        //search for newly scheduled requests
+        auto chunk_begin = std::begin(next_scheduled);
+        auto chunk_end = std::end(next_scheduled);
+        while (chunk_begin != chunk_end) 
+        {
+            std::tie(chunk_begin, std::ignore) = std::mismatch(chunk_begin, chunk_end, std::cbegin(last_scheduled));
+            //found newly scheduled request
+            if (chunk_begin != chunk_end)
+            {
+               scheduled_order.push_back(std::distance(std::begin(next_scheduled), chunk_begin));
+               completed_reqs++;
+               break;
```

Comment:
This break is breaking out of L42, correct? I think we should allow the loop to continue. Is it possible that two packets might be serviced in the same cycle? If not, we should still probably allow the test to mark all newly scheduled packets, since if such a bug were to occur, the test would loop infinitely (on L30).

Reply:
It should not be possible for a channel to schedule multiple packets at once. Though a scheduler may be physically capable of doing such, physical DRAM would be limited to servicing one request at a time (per channel).

Removing this break caused the code to loop forever. I have tried for the past 3 hours to resolve this issue, but am unable to do so. It appears to be an issue with std::mismatch combined with std::transform and std::optional . Multiple packets are not being scheduled at once (as this would also break the current test, as we only record one schedule per operate() call, and thus we would have missing entries in our record vectors. Thankfully, we can expect this test to fail if we end up scheduling multiple packets at once.

I think additional work needs to be done in ChampSim to dissociate bank availability with data being returned. Data will return from the bank before the bank will be available again for more accesses. The turnaround time for memory accesses are fairly quick, as long as a bank is not accessed multiple times in a row, in which we will have to wait for the bank to be ready again. This is why bank interleaving is so important. Let me know if I am missing something in the code that actually respects this.
Currently, ChampSim doesn't do too bad at modeling this, except for :
1. The first access of each bank should see data return much faster than the bank becoming available again.
2. This time offset will carry forward to each data access to that bank following this (since we need to wait for bank availability).
To fix this:
1. We should try to associate an end-event-cycle with each bank which tells us at which cycle the bank will be available again. This can be tied or replace the valid signal each bank currently has.
2. The same can be done for packets coming through the memory controller, and arrive on the data bus at their end-event-cycle. My understanding of the memory controller is that it cannot issue any additional requests until the memory request (write or read) it is servicing is done and ready to be put onto the data bus. Data can be added to the read/write queues at maximum speed regardless. **This might explain the speedup we are seeing from the FR-FCFS algorithm.**
3. There might be other timings that need observed within the model that we are currently ignoring. I am fairly certain timing breakdowns differ with writes vs. reads as well.